File: byteback-test/test-kotlin-1.8.0/src/main/kotlin/byteback/test/exceptions/PotentialNullDereference.kt
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --npe -o %t.bpl
 */

package byteback.test.exceptions

import byteback.specification.Contract.*
import byteback.specification.Operators.*

class A @Return constructor() {

		@Return
		fun method() {}

}

class PotentialNullDereference {

		@Return
		fun freshTarget() {
				val a: A = A()
				a.method()
		}

		@Return
		fun passedTarget(a: A?) {
				if (a != null) {
						a.method()
				}
		}

		@Return
		fun passedArrayTarget(a: Array<A>?) {
				if (a != null) {
						val a_l: Int = a.size;
				}
		}

		@Behavior
		fun `a is not null`(a: IntArray?): Boolean {
				return neq(a, null);
		}

		@Return(`when` = "a is not null")
		fun constrainedPassedArrayTarget(a: IntArray): Unit {
				val a_1: Int = a[1];
		}

}

/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 5 verified, 0 errors
 */

File: byteback-test/test-scala-2.13.8/src/main/scala/byteback/test/exceptions/PotentialNullDereference.scala
/**
  * RUN: %{byteback} -cp %{jar} -c %{class} --npe -o %t.bpl
  */

package byteback.test.exceptions;

import byteback.specification.Contract._;
import byteback.specification.Special._;
import byteback.specification.Operators._;
import byteback.specification.Operators.{eq => equal};

class A @Return () {

  @Return
  def method(): Unit = {}

}

class PotentialNullDereference {

  @Return
  def freshTarget(): Unit = {
    var a: A = new A()
    a.method()
  }

  @Return
  def passedTarget(a: A): Unit = {
    if (a != null) {
      a.method()
    }
  }

  @Return
  def passedArrayTarget(a: Array[Int]): Unit = {
    if (a != null) {
      val a_l: Int = a.length;
    }
  }

  @Behavior
  def a_is_not_null(a: Array[Int]): Boolean = {
    return neq(a, null);
  }

  @Return(when = "a_is_not_null")
  def constrainedPassedArrayTarget(a: Array[Int]): Unit = {
    val a_1: Int = a(1);
  }

}

/**
  * RUN: %{verify} %t.bpl | filecheck %s
  * CHECK: Boogie program verifier finished with 7 verified, 0 errors
  */
